### PR TITLE
add variable product support to api package

### LINKED ISSUE #28883: Add Variable Product Type functionality to `@woocommerce/api` package
Variable Product Type:
- Listing operation
- Creation of variable product model

The first part of completing this issue is
- [ ] create an empty variable product
- [ ] update any tests that create variable products (not as part of the test, but as part of the setup for the test) to use the API to create the variable products for the test.

### PR DESCRIPTION
**Pull request by rrennick**
### Changes proposed in this Pull Request:

This PR add variable product and product variation support to the api package. This includes the following:

- Updates `products.variable` in the local `default.json` to include product properties for testing the api
- Adds `products.variations`  in the local `default.json` for testing the api
- Adds child model creation support
- Updated package tests for child model creation
- Adds E2E tests for variable product and product variations
- Splits out `AbstractProductData` from `AbstractProduct` where product data is the subset of properties for variations
- Splits out `AbstractAttribute` from `ProductAttribute` to support default & variation attributes
- Fixes issue with grouped products from #29214

Closes #28883 .

### How to test the changes in this Pull Request:

1. Verify CI tests in Github
```
  REST API > Variable Product
    ✓ can create a variable product (195ms)
    ✓ can add variations (503ms)
    ✓ can retrieve a transformed variable product (107ms)
    ✓ can retrieve a transformed product variations (77ms)
    ✓ can delete a variation (4ms)
    ✓ can delete a variable product (3ms)
```
2. Run E2E tests locally
3. Edit `tests/e2e/core-tests/api/variable-product.test.js` and comment out the last two tests which delete the test data
4. Edit `products.variable` and `products.variations` in `tests/e2e/config/default.json`
5. Run `npx wc-e2e test:e2e tests/e2e/specs/rest-api/variable-product.js`
6. Verify that the product was created correctly using the web browser
7. Repeat step 3 to just comment out the product delete
8. Re-run step 5 and verify using the browser that the last variation in default.json was deleted.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

N/A

### CODE REVIEW

File: tests/e2e/api/src/framework/model-repository.ts [hunk 1]
```diff
@@ -189,6 +203,21 @@ export interface CreatesModels< T extends ModelRepositoryParams > {
 	create( properties: Partial< ModelClass< T > > ): Promise< ModelClass< T > >;
 }
 
+/**
+ * An interface for repositories that can create child models.
+ *
+ * @typedef CreatesChildModels
+ * @property {CreateChildFn.<T>} create Creates a model using the repository.
```
Comment (zhongruige) [inline tests/e2e/api/src/framework/model-repository.ts:210]: Minor thought (not necessarily worth changing), but we could clarify this is for creating child models:

```suggestion
 * @property {CreateChildFn.<T>} create Creates a child model using the repository.
```

File: tests/e2e/core-tests/specs/api/variable-product.test.js [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+		it('can create a variable product', async () => {
+			productRepository = VariableProduct.restRepository(client);
+
+			// Check properties of product in the create product response.
+			product = await productRepository.create(defaultVariableProduct);
+			expect(product).toEqual(expect.objectContaining(defaultVariableProduct));
+		});
+
+		it('can add variations', async () => {
```
Comment (zhongruige) [inline tests/e2e/core-tests/specs/api/variable-product.test.js:51]: This test doesn't run any asserts on the created variations, is there any we wanted to add here?

**Review by zhongruige — COMMENTED** [review]
These changes are awesome @rrennick! Thanks for taking the time to build out the API package for the variable products!

**Test Results**
* I ran the full e2e test suite which passed. Notably all the API tests continue to pass as well:

Screen Shot 2021-03-03 at 11 43 47 AM

* I updated my `tests/e2e/config/default.json` file with the following changes:
  * Renamed the product
  * Added a new color option (orange)
  * Added two new sizes
  * Added a new logo
  * Adjusted a variation to use the new color
* Ran the tests and verified it was created correctly in the UI:

Screen Shot 2021-03-03 at 1 45 20 PM

Screen Shot 2021-03-03 at 1 55 03 PM

Screen Shot 2021-03-03 at 1 48 55 PM

* Un-commented out `it('can delete a variation')` and verified the last attribute was removed:

Screen Shot 2021-03-03 at 1 49 40 PM

I only had a couple comments, one is a minor one for a small wording change in a doc block; the other is related to assertions in the `it('can add variations'` test block.

File: tests/e2e/core-tests/specs/api/variable-product.test.js (hunk 2 above)
Comment (rrennick) [reply tests/e2e/core-tests/specs/api/variable-product.test.js:51]: Added an assertion for each variation.

File: tests/e2e/api/src/framework/model-repository.ts (hunk 1 above)
Comment (rrennick) [reply tests/e2e/api/src/framework/model-repository.ts]: Added this.

File: tests/e2e/core-tests/specs/api/variable-product.test.js (hunk 2 above)
Comment (zhongruige) [reply tests/e2e/core-tests/specs/api/variable-product.test.js:51]: Thanks!

**Review by zhongruige — APPROVED** [review]
Changes look great and everything ran for me successfully locally:

```
PASS  ../specs/rest-api/variable-product.js (8.144s)
  REST API > Variable Product
    ✓ can create a variable product (183ms)
    ✓ can add variations (499ms)
    ✓ can retrieve a transformed variable product (98ms)
    ✓ can retrieve transformed product variations (72ms)
    ✓ can delete a variation (4ms)
    ✓ can delete a variable product (3ms)

Test Suites: 1 passed, 1 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        8.193s
Ran all test suites matching /tests\/e2e\/specs\/rest-api\/variable-product.js/i.
Jest exit code: 0
```

Once the pipeline finishes up, I'll go ahead and get this merged into trunk.

**Comment by rrennick** [comment]
> Once the pipeline finishes up, I'll go ahead and get this merged into trunk.

@zhongruige Awesome. Thanks for the review.

**Comment by zhongruige** [comment]
> > Once the pipeline finishes up, I'll go ahead and get this merged into trunk.
> 
> @zhongruige Awesome. Thanks for the review.

You're welcome @rrennick! This has been merged in now 👍

**Pull request merged**